src/routes/snippets.rs

```diff
@@ -77,6 +143,68 @@ pub fn create_snippet(
     create_snippet_impl(&**storage, &snippet, origin.path())
 }
 
+fn split_marker(mut snippets: Vec<Snippet>, limit: usize) -> (Option<String>, Vec<Snippet>) {
+    if snippets.len() > limit {
+        snippets.truncate(limit);
+        (snippets.last().map(|m| m.id.to_owned()), snippets)
+    } else {
+        (None, snippets)
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+#[get("/snippets?<title>&<syntax>&<tag>&<marker>&<limit>")]
+pub fn list_snippets<'o, 'h>(
+    storage: State<Box<dyn Storage>>,
+    origin: &'o Origin,
+    title: Option<String>,
+    syntax: Option<String>,
+    tag: Option<String>,
+    limit: Option<Result<PaginationLimit, ApiError>>,
+    marker: Option<String>,
+    _content_type: NegotiatedContentType,
+    _user: BearerAuth,
+) -> Result<WithHttpHeaders<'h, Output<Vec<Snippet>>>, ApiError> {
+    let mut criteria = ListSnippetsQuery {
+        title,
+        syntax,
+        tags: tag.map(|v| vec![v]),
+        ..Default::default()
+    };
+
+    // Fetch one more record in order to detect if there's a next page, and
+    // generate appropriate Link entry accordingly.
+    let limit = limit
+        .unwrap_or_else(|| Ok(<PaginationLimit as Default>::default()))?
+        .0;
+    criteria.pagination.limit = limit + 1;
+    criteria.pagination.marker = marker;
+
+    let snippets = storage.list(criteria.clone())?;
+    let mut prev_needed = false;
+    let (next_marker, snippets) = split_marker(snippets, limit);
+    let prev_marker = if criteria.pagination.marker.is_some() && !snippets.is_empty() {
+        // In order to generate Link entry for previous page we have no choice
+        // but to issue the query one more time into opposite direction.
+        criteria.pagination.direction = Direction::Asc;
+        criteria.pagination.marker = Some(snippets[0].id.to_owned());
+        let prev_snippets = storage.list(criteria)?;
+        prev_needed = !prev_snippets.is_empty();
+
+        prev_snippets.get(limit).map(|m| m.id.to_owned())
```

Comment:
**n.b.** I also replaced `split_marker` here because it does not work for previous snippets page. Let's think about further improvements later on. Long story short, for previous page we're interested in extracting marker at index `limit`, not `limit - 1`. Unfortunately, `split_marker` extracts marker at `limit - 1`.